processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java

```diff
@@ -112,8 +113,12 @@ public TopNQueryEngine(StupidPool<ByteBuffer> bufferPool)
       // A special TimeExtractionTopNAlgorithm is required, since DimExtractionTopNAlgorithm
       // currently relies on the dimension cardinality to support lexicographic sorting
       topNAlgorithm = new TimeExtractionTopNAlgorithm(capabilities, query);
-    } else if(selector.isHasExtractionFn()) {
-      topNAlgorithm = new DimExtractionTopNAlgorithm(capabilities, query);
+    } else if (selector.isHasExtractionFn()) {
+      if (canApplyExtractionInPost(query)) {
```

Comment:
Not here. This one actually stays. It allows using the pooled algorithm, which is much faster, instead of the dim extraction topn.